README.md
<!-- omit in toc -->
# loopback4-cosmosdb-retry

[![Actions Status][build-badge]][actions]
[![Coverage Status][coveralls-badge]][coveralls]

[![Latest version][npm-version-badge]][npm-package]
[![License][license-badge]][license]
[![Downloads][npm-downloads-badge]][npm-package]
[![Total Downloads][npm-total-downloads-badge]][npm-package]

LoopBack 4 data source mixin to handle database operation retries in case of Azure Cosmos DB request
limit errors caused by exceeding the available [Request Units][request-units].

<!-- omit in toc -->
## Contents

- [Prerequisites](#prerequisites)
- [Installation](#installation)
- [Usage](#usage)
- [Configuration](#configuration)
- [Debug](#debug)
- [Related resources](#related-resources)
- [Contributing](#contributing)
- [License](#license)

## Prerequisites

Some dependencies need to be installed as peer dependencies

```shell
@loopback/repository
```

## Installation

```shell
npm install loopback4-cosmosdb-retry
```

## Usage

The mixin just needs to added to the data source

```ts
import { juggler } from "@loopback/repository";
import { RetryMixin } from "loopback4-cosmosdb-retry";

class CosmosdbDataSource extends RetryMixin(juggler.DataSource) {}
```

## Configuration

The [default values][default-values] can be directly overwritten in the data source class

```ts
class CosmosdbDataSource extends RetryMixin(juggler.DataSource) {
    constructor(args) {
        super(args);

        // do 19 retries after the first request (default: 9)
        this.maxRetries = 19;

        // default delay if no suggested delay is in error response (default: 1000)
        this.retryAfterInMs = 2000;

        // additional delay to add to the suggested delay in error response (default: 0)
        this.retryAfterPaddingInMs = 200;

        // always use fixed retry interval based on retryAfterInMs (default: false)
        this.useFixedRetryInterval = false;
    }
}
```

or by using environment variables

> .env

```sh
MAX_RETRIES=19
RETRY_AFTER_IN_MS=2000
RETRY_AFTER_PADDING_IN_MS=200
USE_FIXED_RETRY_INTERVAL=false
```

**Note:** The values you are setting directly in your code will have precedence over the environment variables.
If you want to be able to set values in your code while also having the options to overwrite with environment variables
you need to manually handle this.

```ts

// ...

    constructor(args) {
        super(args);

        this.maxRetries =  Number(process.env.MAX_RETRIES) ?? 19;
    }

// ...

```

## Debug

To enable debug logs set the `DEBUG` environment variable to `loopback:cosmosdb-retry`, see
[Setting debug strings][lb4-debug-strings] for further details.

## Related resources

- [LoopBack 4 Mixins][lb4-mixins]
- [Request Units in Azure Cosmos DB][request-units]

## Contributing

[![contributions welcome][contributions-welcome-badge]][issues]

## License

This project is licensed under the MIT license. See the [LICENSE](LICENSE) file for more info.

[actions]: https://github.com/nflaig/loopback4-cosmosdb-retry/actions
[license]: https://github.com/nflaig/loopback4-cosmosdb-retry/blob/master/LICENSE
[issues]: https://github.com/nflaig/loopback4-cosmosdb-retry/issues
[coveralls]: https://coveralls.io/github/nflaig/loopback4-cosmosdb-retry?branch=refs/heads/master
[npm-package]: https://www.npmjs.com/package/loopback4-cosmosdb-retry

[build-badge]: https://github.com/nflaig/loopback4-cosmosdb-retry/workflows/build/badge.svg
[coveralls-badge]: https://coveralls.io/repos/github/nflaig/loopback4-cosmosdb-retry/badge.svg?branch=refs/heads/master
[npm-version-badge]: https://img.shields.io/npm/v/loopback4-cosmosdb-retry.svg?style=flat-square
[npm-downloads-badge]: https://img.shields.io/npm/dw/loopback4-cosmosdb-retry.svg?label=Downloads&style=flat-square&color=blue
[npm-total-downloads-badge]: https://img.shields.io/npm/dt/loopback4-cosmosdb-retry.svg?label=Total%20Downloads&style=flat-square&color=blue
[license-badge]: https://img.shields.io/github/license/nflaig/loopback4-cosmosdb-retry.svg?color=blue&label=License&style=flat-square
[contributions-welcome-badge]: https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat

[request-units]: http://aka.ms/cosmosdb-error-429
[lb4-mixins]: https://loopback.io/doc/en/lb4/Mixin.html
[lb4-debug-strings]: https://loopback.io/doc/en/lb4/Setting-debug-strings.html
[default-values]: https://github.com/nflaig/loopback4-cosmosdb-retry/blob/master/src/retry.ts#L4


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## [1.2.1](https://github.com/nflaig/loopback4-cosmosdb-retry/compare/v1.2.0...v1.2.1) (2022-01-13)

### Improvements

* Update dependencies




## [1.2.0](https://github.com/nflaig/loopback4-cosmosdb-retry/compare/v1.1.0...v1.2.0) (2021-01-24)

### Features

* Add debug logs of retry attempts




## [1.1.0](https://github.com/nflaig/loopback4-cosmosdb-retry/compare/v1.0.0...v1.1.0) (2020-09-21)

### Documentation

* Add table of contents to README




## 1.0.0 (2020-09-05)

Initial version


src/__tests__/helpers/index.ts
export * from "./application.helper";
export * from "./datasource.helper";
export * from "./repository.helper";


src/__tests__/unit/retry.unit.ts
import { expect } from "@loopback/testlab";
import {
    shouldRetry,
    getDelayInMs,
    getRetryAfterMs,
    getErrorMessage,
    isError,
    isErrorArray
} from "../../retry";
import {
    cosmosdbRequestLimitError,
    cosmosdbRetryAfterMs,
    cosmosdbRequestLimitMessage
} from "../fixtures/data";

describe("Retry (unit)", () => {
    class ErrorWithInvalidCode extends Error {
        code = "shouldBeANumber";
    }

    const unknownError = new Error("something failed");
    const errorWithInvalidCode = new ErrorWithInvalidCode("code should be a number");

    describe("shouldRetry()", () => {
        it("should return true if a Cosmos DB request limit error is provided", () => {
            const retry = shouldRetry(cosmosdbRequestLimitError);

            expect(retry).to.equal(true);
        });

        it("should return true if a Cosmos DB request limit error is provided as an array", () => {
            const retry = shouldRetry([cosmosdbRequestLimitError, cosmosdbRequestLimitError]);

            expect(retry).to.equal(true);
        });

        it("should return false if an unknown error is provided", () => {
            const retry = shouldRetry(unknownError);

            expect(retry).to.equal(false);
        });
    });

    describe("getDelayInMs()", () => {
        it("should return the delay based on the RetryAfterMs value in the Cosmos DB request limit error", () => {
            const delay = getDelayInMs(cosmosdbRequestLimitError);

            expect(delay).to.equal(cosmosdbRetryAfterMs);
        });

        it("should return undefined if an array of errors is provided", () => {
            const delay = getDelayInMs([cosmosdbRequestLimitError, cosmosdbRequestLimitError]);

            expect(delay).to.be.undefined();
        });

        it("should return undefined if an unknown error is provided", () => {
            const delay = getDelayInMs(unknownError);

            expect(delay).to.be.undefined();
        });
    });

    describe("getRetryAfterMs()", () => {
        it("should retrieve and return the RetryAfterMs value from the Cosmos DB request limit error message", () => {
            const retryAfterInMs = getRetryAfterMs(cosmosdbRequestLimitMessage);

            expect(retryAfterInMs).to.equal(cosmosdbRetryAfterMs);
        });

        it("should add the padding to the returned RetryAfterMs value", () => {
            const retryAfterPaddingInMs = 10;
            const retryAfterInMs = getRetryAfterMs(
                cosmosdbRequestLimitMessage,
                retryAfterPaddingInMs
            );

            expect(retryAfterInMs).to.equal(cosmosdbRetryAfterMs + retryAfterPaddingInMs);
        });

        it("should return undefined if an unknown error message is provided", () => {
            const retryAfterInMs = getRetryAfterMs(unknownError.message);

            expect(retryAfterInMs).to.be.undefined();
        });

        it("should return undefined if an unknown error message is not a string", () => {
            const retryAfterInMs = getRetryAfterMs(<string>{});

            expect(retryAfterInMs).to.be.undefined();
        });

        it("should return undefined if RetryAfterMs is not a number", () => {
            const retryAfterInMs = getRetryAfterMs("RetryAfterMs=DefinitelyNotANumber");

            expect(retryAfterInMs).to.be.undefined();
        });
    });

    describe("getErrorMessage()", () => {
        it("should return the message of the provided error", () => {
            const message = getErrorMessage(cosmosdbRequestLimitError);

            expect(message).to.equal(cosmosdbRequestLimitError.message);
        });

        it("should return the message if the error is provided as an array", () => {
            const message = getErrorMessage([cosmosdbRequestLimitError]);

            expect(message).to.equal(cosmosdbRequestLimitError.message);
        });

        it("should return undefined if an unknown error is provided", () => {
            const message = getErrorMessage(unknownError);

            expect(message).to.be.undefined();
        });

        it("should return undefined if an unknown error is provided as an array", () => {
            const message = getDelayInMs([unknownError]);

            expect(message).to.be.undefined();
        });

        it("should return undefined if array is empty", () => {
            const message = getDelayInMs([]);

            expect(message).to.be.undefined();
        });
    });

    describe("isError()", () => {
        it("should return true if the provided error is a proper error response", () => {
            const isErrorResponse = isError(cosmosdbRequestLimitError);

            expect(isErrorResponse).to.equal(true);
        });

        it("should return false if an unknown error is provided", () => {
            const isErrorResponse = isError(unknownError);

            expect(isErrorResponse).to.equal(false);
        });

        it("should return false if an error with an invalid code type is provided", () => {
            const isErrorResponse = isError(errorWithInvalidCode);

            expect(isErrorResponse).to.equal(false);
        });

        it("should return false if the provided error is undefined", () => {
            const isErrorResponse = isError(undefined);

            expect(isErrorResponse).to.equal(false);
        });
    });

    describe("isErrorArray()", () => {
        it("should return true if the provided error is a proper error response as array", () => {
            const isErrorResponseArray = isErrorArray([cosmosdbRequestLimitError]);

            expect(isErrorResponseArray).to.equal(true);
        });

        it("should return false if an unknown error is provided as array", () => {
            const isErrorResponseArray = isErrorArray([unknownError]);

            expect(isErrorResponseArray).to.equal(false);
        });

        it("should return false if an error with an invalid code type is provided as array", () => {
            const isErrorResponseArray = isErrorArray([errorWithInvalidCode]);

            expect(isErrorResponseArray).to.equal(false);
        });

        it("should return false if array is empty", () => {
            const isErrorResponseArray = isErrorArray([]);

            expect(isErrorResponseArray).to.equal(false);
        });

        it("should return false if the provided error is not an array", () => {
            const isErrorResponseArray = isErrorArray(cosmosdbRequestLimitError);

            expect(isErrorResponseArray).to.equal(false);
        });
    });
});


src/__tests__/unit/utils.unit.ts
import { expect } from "@loopback/testlab";
import { getEnvNumber, getEnvBoolean, pluralize } from "../../utils";

describe("Utils (unit)", () => {
    describe("pluralize()", () => {
        const noun = "test";

        it("should pluralize the given noun and display the count", () => {
            const count = 2;

            const pluralized = pluralize(noun, count);

            expect(pluralized).to.equal(`${count} ${noun}s`);
        });

        it("should not pluralize the given noun if the count is one", () => {
            const count = 1;

            const pluralized = pluralize(noun, count);

            expect(pluralized).to.equal(`${count} ${noun}`);
        });

        it("should not display the count if the prefixNumber is set to false", () => {
            const count = 1;

            const pluralized = pluralize(noun, count, false);

            expect(pluralized).to.equal(noun);
        });

        it("should append a different suffix if specified", () => {
            const count = 2;

            const differentSuffix = "differentSuffix";
            const pluralized = pluralize(noun, count, false, differentSuffix);

            expect(pluralized).to.equal(noun + differentSuffix);
        });
    });

    describe("getEnvNumber()", () => {
        const key = "TEST_NUMBER";
        const defaultValue = 123;
        const fallbackValue = 456;

        beforeEach(() => {
            setEnvNumber(defaultValue);
        });

        it("should return the environment variable as a number", () => {
            const testNumber = getEnvNumber(key, fallbackValue);

            expect(testNumber).to.equal(defaultValue);
        });

        it("should return the fallback value if the environment variable is not set", () => {
            unsetEnvNumber();

            const testNumber = getEnvNumber(key, fallbackValue);

            expect(testNumber).to.equal(fallbackValue);
        });

        it("should return the fallback value if the environment variable is not a number", () => {
            setEnvNumber(<number>(<unknown>"abc"));

            const testNumber = getEnvNumber(key, fallbackValue);

            expect(testNumber).to.equal(fallbackValue);
        });

        function setEnvNumber(value: number) {
            process.env[key] = value.toString();
        }

        function unsetEnvNumber() {
            delete process.env[key];
        }
    });

    describe("getEnvBoolean()", () => {
        const key = "TEST_BOOLEAN";
        const defaultValue = true;
        const fallbackValue = false;

        beforeEach(() => {
            setEnvBoolean(defaultValue);
        });

        it("should return the environment variable as a boolean", () => {
            const testBoolean = getEnvBoolean(key, fallbackValue);

            expect(testBoolean).to.equal(defaultValue);
        });

        it("should return the fallback value if the environment variable is not set", () => {
            unsetEnvBoolean();

            const testBoolean = getEnvBoolean(key, fallbackValue);

            expect(testBoolean).to.equal(fallbackValue);
        });

        it("should return the fallback value if the environment variable is not a boolean", () => {
            setEnvBoolean(<boolean>(<unknown>"abc"));

            const testBoolean = getEnvBoolean(key, fallbackValue);

            expect(testBoolean).to.equal(fallbackValue);
        });

        function setEnvBoolean(value: boolean) {
            process.env[key] = value.toString();
        }

        function unsetEnvBoolean() {
            delete process.env[key];
        }
    });
});


.eslintrc.js
module.exports = {
    extends: ["eslint:recommended", "@loopback/eslint-config"],
    overrides: [
        {
            files: ["src/__tests__/fixtures/models/**"],
            rules: { indent: "off" }
        }
    ],
    parserOptions: {
        ecmaVersion: 9,
        sourceType: "module"
    },
    globals: {
        require: true
    },
    env: {
        es6: true
    },
    rules: {
        "no-unused-vars": 2,
        "global-require": 1,
        "no-empty": 2,
        quotes: [0],
        "no-console": 2,
        "no-undef": 2,
        "no-const-assign": 2,
        "prefer-const": 2,
        "func-call-spacing": 2,
        "no-undef-init": 2,
        indent: [
            2,
            4,
            {
                FunctionExpression: { parameters: "first" },
                CallExpression: { arguments: "first" },
                SwitchCase: 1,
                ObjectExpression: "first",
                ArrayExpression: "first"
            }
        ],
        curly: [2, "multi-line"],
        "object-curly-spacing": [2, "always"],
        "comma-dangle": [2, "never"],
        "max-len": [2, 150],
        "max-params": [1, 4],
        "max-lines": [1, 500],
        "max-depth": [1, 3],
        "max-nested-callbacks": [1, 3],
        "@typescript-eslint/no-explicit-any": "off",
        "no-restricted-modules": [
            2,
            {
                paths: [
                    {
                        name: "winston",
                        message: "Please use the wrapper instead."
                    },
                    {
                        name: "fs-extra",
                        message: "Please use the 'fs' module instead."
                    },
                    {
                        name: "async",
                        message: "Please use the native 'Promise' instead."
                    },
                    {
                        name: "mkdirp",
                        message: "Please use the 'fs' module instead."
                    }
                ]
            }
        ]
    }
};


src/mixins/retry.datasource.mixin.ts
import { juggler } from "@loopback/repository";
import {
    MAX_RETRIES,
    RETRY_AFTER_IN_MS,
    RETRY_AFTER_PADDING_IN_MS,
    USE_FIXED_RETRY_INTERVAL,
    shouldRetry,
    getDelayInMs,
    TooManyRequestsError,
    logRetry,
    logTooManyRequests
} from "../retry";
import { MixinTarget } from "../types";

export function RetryMixin<T extends MixinTarget<juggler.DataSource>>(dataSourceClass: T) {
    class RetryDataSource extends dataSourceClass {
        maxRetries: number;
        retryAfterInMs: number;
        retryAfterPaddingInMs: number;
        useFixedRetryInterval: boolean;

        constructor(...args: any[]) {
            super(...args);

            this.maxRetries = MAX_RETRIES;
            this.retryAfterInMs = RETRY_AFTER_IN_MS;
            this.retryAfterPaddingInMs = RETRY_AFTER_PADDING_IN_MS;
            this.useFixedRetryInterval = USE_FIXED_RETRY_INTERVAL;

            const connector = this.connector!;
            // save the original `execute` implementation
            connector._originalExecute = connector.execute;
            // inject custom version of `execute`
            (connector.execute as (...params: any[]) => any) = (...params: unknown[]) => {
                // connectors are still callback-based, LB4 type definitions are incorrect
                const callback = params.pop() as Function;
                const [collection, method] = params as [string, string];

                let attempt = 1;

                const retryCallback = (error: unknown, ...results: unknown[]) => {
                    if (!error) {
                        // no error occurred, operation was successful
                        callback(error, ...results);
                        return;
                    }

                    if (!shouldRetry(error)) {
                        // unknown error, report it back to the caller
                        callback(error, ...results);
                        return;
                    }

                    const delay = this.useFixedRetryInterval
                        ? this.retryAfterInMs
                        : getDelayInMs(error, this.retryAfterPaddingInMs) ?? this.retryAfterInMs;

                    logRetry(attempt, delay, this.maxRetries, method, collection);

                    if (attempt <= this.maxRetries) {
                        attempt++;
                        setTimeout(
                            () => connector._originalExecute(...params, retryCallback),
                            delay
                        );
                    } else {
                        logTooManyRequests(error);
                        callback(new TooManyRequestsError(delay), ...results);
                    }
                };

                // forward the arguments provided by the caller
                // but use custom callback to intercept the outcome
                connector._originalExecute(...params, retryCallback);
            };
        }
    }
    return RetryDataSource;
}


src/retry.ts
import debugFactory from "debug";
import { ErrorResponse } from "./types";
import { getEnvNumber, getEnvBoolean, pluralize } from "./utils";

const debug = debugFactory("loopback:cosmosdb-retry");

export const MAX_RETRIES = getEnvNumber("MAX_RETRIES", 9);
export const RETRY_AFTER_IN_MS = getEnvNumber("RETRY_AFTER_IN_MS", 1000);
export const RETRY_AFTER_PADDING_IN_MS = getEnvNumber("RETRY_AFTER_PADDING_IN_MS", 0);
export const USE_FIXED_RETRY_INTERVAL = getEnvBoolean("USE_FIXED_RETRY_INTERVAL", false);

export class TooManyRequestsError extends Error {
    code?: string;
    statusCode?: number;
    details: {
        retryAfterInMs?: number;
    };

    constructor(retryAfterInMs?: number) {
        super("Too many requests, please try again later.");
        this.name = this.constructor.name;
        this.code = "TOO_MANY_REQUESTS";
        this.statusCode = 429;
        this.details = { retryAfterInMs };
    }
}

export function shouldRetry(error?: any): boolean {
    const retryErrorCode = 16500;

    if (isError(error)) {
        return error.code === retryErrorCode;
    } else if (isErrorArray(error)) {
        return error[0].code === retryErrorCode;
    } else {
        return false;
    }
}

// eslint-disable-next-line max-params
export function logRetry(
    attempt: number,
    delayInMs: number,
    maxRetries: number,
    method: string,
    collection: string
) {
    const attempts = pluralize("time", attempt);
    const baseMessage = `Database operation ${method} on ${collection} collection has failed ${attempts}.`;
    const retryMessage = `${baseMessage} Retrying after ${delayInMs}ms...`;
    const maxAmountMessage = `${baseMessage} Maximum amount of retries has been reached.`;

    if (attempt <= maxRetries) {
        debug(retryMessage);
    } else {
        debug(maxAmountMessage);
    }
}

export function logTooManyRequests(error?: any) {
    debug(getErrorMessage(error));
}

export function getDelayInMs(error?: any, retryAfterPaddingInMs = 0): number | undefined {
    if (isError(error)) {
        // get suggested wait period from error response
        return getRetryAfterMs(error.message, retryAfterPaddingInMs);
    } else if (isErrorArray(error)) {
        // use default wait period
        return;
    }
}

export function getRetryAfterMs(
    errorMessage: string,
    retryAfterPaddingInMs = 0
): number | undefined {
    if (typeof errorMessage !== "string") return;

    const matcher = /RetryAfterMs=[0-9]*/;

    const result = errorMessage.match(matcher);

    if (!result) return;

    const retryAfterInMs = result[0].split("=")[1];

    return retryAfterInMs ? Number(retryAfterInMs) + retryAfterPaddingInMs : undefined;
}

export function getErrorMessage(error?: any): string | undefined {
    if (isError(error)) {
        return error.message;
    } else if (isErrorArray(error)) {
        return error[0].message;
    }
}

export function isError(error?: any): error is ErrorResponse {
    return typeof error?.code === "number";
}

export function isErrorArray(error?: any): error is ErrorResponse[] {
    return Array.isArray(error) && typeof error[0]?.code === "number";
}
